File: ast/src/testing/kotlin/app/src/main/java/com/kotlintestapp/MethodChainFactory.kt
package com.kotlintestapp
// Tests free-function return-type seeding (fn_returns) for Kotlin.
// makeWidget() returns Widget; fn_returns seeds the return type so that
// `val w = makeWidget()` binds w → Widget, enabling w.render() to
// resolve to Widget::render via find_method_in_class.
//
// @ast node: Class "Widget"
// @ast node: Function "render"
// @ast node: Function "makeWidget"
// @ast node: Function "factoryPipeline"
// @ast edge: Calls -> Function "makeWidget" "MethodChainFactory.kt"
// @ast edge: Calls -> Function "render" "MethodChainFactory.kt"
// @ast node: Import "import-imports-srctestingkotlinappsrcmainjavacomkotlintestappmethodchainfactorykt-0"

class Widget {
    fun render() {}
}

fun makeWidget(): Widget = Widget()

fun factoryPipeline() {
    val w = makeWidget()
    w.render()
}

File: ast/src/testing/c/systems-programming/method_chain_factory.c
// Tests free-function call resolution via the C registry.
//
// create_pool() and use_pool() are free functions.  The resolver
// seeds dir_fns so that the call to create_pool() inside use_pool()
// resolves via directory-scoped function lookup.
//
// @ast node: Class "SimplePool"
// @ast node: Function "create_pool"
// @ast node: Function "use_pool"
// @ast edge: Calls -> Function "create_pool" "method_chain_factory.c"
#include <stdlib.h>

typedef struct {
    int size;
} SimplePool;

SimplePool* create_pool(int size) {
    SimplePool* p = (SimplePool*)malloc(sizeof(SimplePool));
    p->size = size;
    return p;
}

void use_pool() {
    SimplePool* pool = create_pool(64);
    pool->size = 128;
}
